### PR TITLE
isZeroReport utility function for Analytics

### PR DESCRIPTION
**Pull request by benbowler**
## Summary

Add a new utility for Analytics widgets to detect empty data and show the empty component in this case.

Addresses issue #2243 

## Relevant technical choices

I've added the new helper to this file `assets/js/modules/analytics/util/is-zero-report.js` with tests here `assets/js/modules/analytics/util/is-zero-report.test.js` before applying it to each template that loads this report type.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/analytics/util/is-zero-report.test.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+				data: {
+					rows: [
+						{}, {}, {},
+					],
+					totals: [
+						{
+							values: [
+								'928',
+								'928',
+							],
```
Comment (aaemnnosttv) [inline assets/js/modules/analytics/util/is-zero-report.test.js:76]: Please add another case that has values with `0` and a positive number to show that it doesn't require all values to be non-zero.

File: assets/js/modules/analytics/util/is-zero-report.test.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+			{
+				data: {
+					rows: [
+						{}, {}, {},
+					],
+					totals: [
+						{
+							values: [
+								'928',
+								'928',
```
Comment (aaemnnosttv) [inline assets/js/modules/analytics/util/is-zero-report.test.js:75]: Bit of a nitpick here but let's make it a little more clear that these are dummy values and that there is no significance to them being the same number.
```suggestion
								'123',
								'234',
```

File: assets/js/modules/analytics/util/is-zero-report.test.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+						{}, {}, {},
+					],
+					totals: [
+						{
+							values: [
+								'0',
+								'0',
+							],
+						},
+					],
```
Comment (aaemnnosttv) [inline assets/js/modules/analytics/util/is-zero-report.test.js:56]: Not to say that we should mirror our real/expected usage in tests (we shouldn't) but we should cover the cases that we use which aren't all covered here yet.
* Multiple items in `totals` (most of the time we just have 1, but some have 2, depending on the number of date ranges requested – 2 will be more common in the future)  
This also relates to my previous comment about whether or not we should consider values from other totals or only the first one (pending input from @felixarntz) but is worth including here either way.
* A single value in the list of `values` (depends on the number of metrics requested, but we usually only request one per report it seems)

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @benbowler – this is off to a good start, just a few things to address – mostly little things and a few tests to add/tweak. There's one comment where I've asked for clarification from @felixarntz but that shouldn't block the rest of it from being addressed.

File: assets/js/modules/analytics/util/is-zero-report.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+export function isZeroReport( report ) {
+	if ( report === undefined ) {
+		return undefined;
+	}
+
+	if ( ! report?.[ 0 ]?.data?.rows?.length || ! report?.[ 0 ]?.data?.totals?.[ 0 ]?.values?.length ) {
+		return true;
+	}
+
+	const sumOfTotals = report[ 0 ].data.totals.reduce( function( totalSum, values ) {
```
[the remark this answers is not in the document]
Comment (benbowler) [reply assets/js/modules/analytics/util/is-zero-report.js:36]: I've updated to check the first totals result with 'Array.some' now based on the AB and will wait for input from @felixarntz on expanding to all items.

File: assets/js/modules/analytics/util/is-zero-report.test.js (hunk 3 above)
Comment (benbowler) [reply assets/js/modules/analytics/util/is-zero-report.test.js:56]: I've added in tests for multiple totals objects and single, multiple and mixed 0 and non-0 values.

File: assets/js/modules/analytics/util/is-zero-report.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixarntz) [reply assets/js/modules/analytics/util/is-zero-report.js]: @benbowler @aaemnnosttv Good point, let's expand the check to cover all `report[ 0 ].data.totals`, not just the first one (there will only be a maximum of two date ranges per how the API works fyi). If the previous date range has data, it should indeed be an indicator that this is not a "zero data" state.

[the remark this answers is not in the document]
Comment (benbowler) [reply assets/js/modules/analytics/util/is-zero-report.js]: Thanks Felix, I'll get this updated and ready for CR.

[the remark this answers is not in the document]
Comment (benbowler) [reply assets/js/modules/analytics/util/is-zero-report.js]: I've updated the function to test all totals values and simplified it to a single line return.

**Comment by benbowler** [comment]
Updated based on feedback.

File: assets/js/modules/analytics/util/is-zero-report.test.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+		[ 'a string', 'test' ],
+		[ 'an empty object', {} ],
+		[ 'an object without rows or totals', [ { data: {} } ] ],
+		[ 'an object with empty rows', [ { data: { rows: [] } } ] ],
+		[ 'an object with empty totals', [ { data: { totals: [] } } ] ],
+	] )( 'returns true when %s is passed', ( _, report ) => {
+		expect( isZeroReport( report ) ).toBe( true );
+	} );
+
+	it( 'returns true id all totals are 0', () => {
```
Comment (aaemnnosttv) [inline assets/js/modules/analytics/util/is-zero-report.test.js:42]: Oops!
```suggestion
	it( 'returns true if all totals are 0', () => {
```

File: assets/js/modules/analytics/util/is-zero-report.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+
+/**
+ * Checks whether the report data is empty or not.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object} report Report data object.
+ * @return {boolean | undefined} Returns undefined if in the loading state, true if the report has no data or missing data, otherwise false.
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/analytics/util/is-zero-report.js:25]: This is yet to be addressed.

File: assets/js/modules/analytics/util/is-zero-report.test.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+								'123',
+							],
+						},
+					],
+				},
+			},
+		];
+
+		expect( isZeroReport( reportSingleTotal ) ).toBe( false );
+	} );
```
Comment (aaemnnosttv) [inline assets/js/modules/analytics/util/is-zero-report.test.js:194]: Please split the assertions in the last two tests here into separate tests. That is, those two currently have 3 expect calls which work fine as-is, but would be a bit more useful in the case of a failure to have separately with descriptive names for each.

**Review by aaemnnosttv — COMMENTED** [review]
This one is super close @benbowler – just a few final small things left to address and then this should be good to go 👍

File: assets/js/modules/analytics/util/is-zero-report.test.js (hunk 5 above)
Comment (benbowler) [reply assets/js/modules/analytics/util/is-zero-report.test.js]: Done.

File: assets/js/modules/analytics/util/is-zero-report.test.js (hunk 7 above)
Comment (benbowler) [reply assets/js/modules/analytics/util/is-zero-report.test.js:200]: Both done and pushed.

Comment (aaemnnosttv) [reply assets/js/modules/analytics/util/is-zero-report.test.js:200]: Thanks @benbowler – it looks like you only split the last test, can you do it for the previous one as well? Apologies if this wasn't clear from my last comment.

**Review by aaemnnosttv — APPROVED** [review]
LGTM 👍

**Pull request merged**